File: workbook/ch07/sec7.3/paradigms/imp/fortran/fortran.py
import re
import sys

import re
from typing import Dict, List, Tuple, Union, Optional, Any, Pattern, Match


class FortranLikeInterpreter:
    
    def __init__(self) -> None:
        self.variables: Dict[str, Union[int, float]] = {}
        self.labels: Dict[str, int] = {}
        self.program: List[Tuple] = []
        self.pc: int = 0  # program counter
        self.running: bool = False
        
    def parse_line(self, line: str) -> Optional[Tuple]:
        line = line.strip().upper()
        if not line or line.startswith('!'):
            return None
            
        # label parsing
        label_match: Optional[Match] = re.match(r'^(\d+)\s+(.*)', line)
        if label_match:
            label, rest = label_match.groups()
            self.labels[label] = len(self.program)
            line = rest.strip()
 
        # operation type
        if ':=' in line:
            var, expr = line.split(':=', 1)
            return ('assign', var.strip(), expr.strip())
        elif line.startswith('GOTO'):
            return ('goto', line[4:].strip())
        elif line.startswith('IF'):
            match: Optional[Match] = re.match(r'IF\s*\((.*?)\)\s*GOTO\s*(\d+)', line)
            if match:
                return ('if_goto', *match.groups())
        elif line.startswith('PRINT'):
            items = line[5:].strip(' *').split(',')
            return ('print', [item.strip() for item in items])
        elif line == 'END':
            return ('end',)
        return None

    def tokenize(self, expr: str) -> List[str]:
        tokens: List[str] = []
        current: str = ''
        for char in expr:
            if char in '+-*/()':
                if current:
                    tokens.append(current)
                    current = ''
                tokens.append(char)
            elif char.isspace():
                if current:
                    tokens.append(current)
                    current = ''
            else:
                current += char
        if current:
            tokens.append(current)
        return tokens

    def shunting_yard(self, tokens: List[str]) -> List[Union[str, int, float]]:
        precedence: Dict[str, int] = {'+': 1, '-': 1, '*': 2, '/': 2}
        output: List[Union[str, int, float]] = []
        stack: List[str] = []
        
        for token in tokens:
            if token.replace('.', '', 1).isdigit():
                num: Union[int, float] = float(token) if '.' in token else int(token)
                output.append(num)
            elif token in precedence:
                while stack and stack[-1] != '(' and precedence[token] <= precedence.get(stack[-1], 0):
                    output.append(stack.pop())
                stack.append(token)
            elif token == '(':
                stack.append(token)
            elif token == ')':
                while stack and stack[-1] != '(':
                    output.append(stack.pop())
                if stack:  # Guard against empty stack
                    stack.pop()
                else:
                    raise ValueError(f"Mismatched parentheses in expression")
            else:
                raise ValueError(f"Invalid token: {token}")
                
        while stack:
            if stack[-1] == '(':
                raise ValueError(f"Mismatched parentheses in expression")
            output.append(stack.pop())
            
        return output

    def evaluate_postfix(self, postfix: List[Union[str, int, float]]) -> Union[int, float]:
        stack: List[Union[int, float]] = []
        
        for token in postfix:
            if isinstance(token, (int, float)):
                stack.append(token)
            else:
                if len(stack) < 2:
                    raise ValueError(f"Invalid expression: not enough operands")
                
                b = stack.pop()
                a = stack.pop()
                
                if token == '+': 
                    stack.append(a + b)
                elif token == '-': 
                    stack.append(a - b)
                elif token == '*': 
                    stack.append(a * b)
                elif token == '/': 
                    if b == 0:
                        raise ValueError("Division by zero")
                    stack.append(a / b)
                    
        if len(stack) != 1:
            raise ValueError(f"Invalid expression: too many operands")
            
        return stack[0]

    def evaluate_expression(self, expr: str) -> Union[int, float]:
        # sub variables with longest names first to prevent partial matches
        for var in sorted(self.variables.keys(), key=lambda x: -len(x)):
            expr = expr.replace(var, str(self.variables[var]))

        try:
            tokens = self.tokenize(expr)
            postfix = self.shunting_yard(tokens)
            return self.evaluate_postfix(postfix)
        except Exception as e:
            raise ValueError(f"Error evaluating expression: {expr}. {str(e)}")

    def evaluate_condition(self, cond: str) -> bool:
        # convert Fortran-style operators
        cond = cond.replace('.EQ.', '==').replace('.NE.', '!=')
        cond = cond.replace('.LT.', '<').replace('.LE.', '<=')
        cond = cond.replace('.GT.', '>').replace('.GE.', '>=')

        # split into components
        operators = ['!=', '==', '<=', '>=', '<', '>', '=']
        for op in operators:
            if op in cond:
                left, right = cond.split(op, 1)
                left_val = self.evaluate_expression(left.strip())
                right_val = self.evaluate_expression(right.strip())
                
                if op == '==': return left_val == right_val
                elif op == '!=': return left_val != right_val
                elif op == '<': return left_val < right_val
                elif op == '<=': return left_val <= right_val
                elif op == '>': return left_val > right_val
                elif op == '>=': return left_val >= right_val
                elif op == '=': return left_val == right_val
                
        raise ValueError(f"Invalid condition: {cond}")

    def execute(self, instruction: Tuple) -> None:
        op = instruction[0]
        
        if op == 'assign':
            _, var, expr = instruction
            self.variables[var] = self.evaluate_expression(expr)
            self.pc += 1
            
        elif op == 'goto':
            _, label = instruction
            if label in self.labels:
                self.pc = self.labels[label]
            else:
                raise ValueError(f"Undefined label: {label}")
                
        elif op == 'if_goto':
            _, condition, label = instruction
            if self.evaluate_condition(condition):
                if label in self.labels:
                    self.pc = self.labels[label]
                else:
                    raise ValueError(f"Undefined label: {label}")
            else:
                self.pc += 1
                
        elif op == 'print':
            _, items = instruction
            output: List[str] = []
            for item in items:
                if item in self.variables:
                    output.append(str(self.variables[item]))
                else:
                    output.append(item)
            print(' '.join(output))
            self.pc += 1
            
        elif op == 'end':
            self.running = False
                
    def run(self) -> None:
        self.pc = 0
        self.running = True
        
        while self.running and self.pc < len(self.program):
            instruction = self.program[self.pc]
            self.execute(instruction)

    def load_program(self, code: str) -> None:
        self.program = []
        self.labels = {}
        
        for line in code.split('\n'):
            instruction = self.parse_line(line)
            if instruction:
                self.program.append(instruction)

    def load_from_file(self, filename: str) -> None:
        with open(filename, 'r') as f:
            code = f.read()
        self.load_program(code)

# example
if __name__ == "__main__":
    if len(sys.argv) > 1:
        # program from file
        interpreter = FortranLikeInterpreter()
        try:
            interpreter.load_from_file(sys.argv[1])
            interpreter.run()
        except FileNotFoundError:
            print(f"Error: File '{sys.argv[1]}' not found")
    else:
        # built-in sample
        interpreter = FortranLikeInterpreter()
        program = """
        ! Simple factorial calculation
           N := 5
           FACT := 1
        10 IF (N .GT. 1) GOTO 20
           PRINT *, "FACTORIAL IS", FACT
           GOTO 30
        20 FACT := FACT * N
           N := N - 1
           GOTO 10
        30 END
        """
        interpreter.load_program(program)
        interpreter.run()

File: workbook/ch07/sec7.3/paradigms/imp/fortran/fortran.c
// fortran_interpreter.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_PROGRAM_LINES 1000
#define MAX_VARIABLES 100
#define MAX_LINE_LENGTH 256

typedef enum { ASSIGN, GOTO, IFGOTO, PRINT, END } InstructionType;

typedef struct {
    InstructionType type;
    char var1[32];
    char var2[32];
    char op[4];
    char label[10];
    char expr[128];
} Instruction;

typedef struct {
    char name[32];
    double value;
} Variable;

Instruction program[MAX_PROGRAM_LINES];
int program_size = 0;
int labels[MAX_PROGRAM_LINES]; // label number -> program index

Variable variables[MAX_VARIABLES];
int variable_count = 0;

int find_variable(const char* name) {
    for (int i = 0; i < variable_count; i++) {
        if (strcmp(variables[i].name, name) == 0)
            return i;
    }
    return -1;
}

void set_variable(const char* name, double value) {
    int idx = find_variable(name);
    if (idx >= 0) {
        variables[idx].value = value;
    } else {
        if (variable_count >= MAX_VARIABLES) {
            fprintf(stderr, "Error: Too many variables.\n");
            exit(1);
        }
        strncpy(variables[variable_count].name, name, sizeof(variables[variable_count].name) - 1);
        variables[variable_count].name[sizeof(variables[variable_count].name) - 1] = '\0';
        variables[variable_count].value = value;
        variable_count++;
    }
}

double get_variable(const char* name) {
    int idx = find_variable(name);
    if (idx >= 0)
        return variables[idx].value;
    fprintf(stderr, "Error: Undefined variable '%s'\n", name);
    exit(1);
}

double eval_expr(const char* expr);

void parse_line(char* line) {
    char* p = line;
    while (isspace(*p)) p++;

    // Check for label
    if (isdigit(*p)) {
        char label_str[10];
        int i = 0;
        while (isdigit(*p) && i < (int)(sizeof(label_str) - 1)) {
            label_str[i++] = *p++;
        }
        label_str[i] = '\0';
        while (isspace(*p)) p++;
        int label_num = atoi(label_str);
        if (label_num < 0 || label_num >= MAX_PROGRAM_LINES) {
            fprintf(stderr, "Error: Invalid label number %d\n", label_num);
            exit(1);
        }
        labels[label_num] = program_size;
    }

    if (strncmp(p, "IF", 2) == 0) {
        p += 2;
        while (isspace(*p)) p++;
        if (*p != '(') {
            fprintf(stderr, "Error: Expected '(' after IF\n");
            exit(1);
        }
        p++;
        char condition[64];
        int i = 0;
        while (*p && *p != ')' && i < (int)(sizeof(condition) - 1)) {
            condition[i++] = *p++;
        }
        condition[i] = '\0';
        if (*p != ')') {
            fprintf(stderr, "Error: Expected ')' after condition\n");
            exit(1);
        }
        p++;
        while (isspace(*p)) p++;
        if (strncmp(p, "GOTO", 4) != 0) {
            fprintf(stderr, "Error: Expected GOTO after IF (...)\n");
            exit(1);
        }
        p += 4;
        while (isspace(*p)) p++;
        char label_target[10];
        sscanf(p, "%9s", label_target);

        Instruction instr = {0};
        instr.type = IFGOTO;
        strncpy(instr.expr, condition, sizeof(instr.expr) - 1);
        strncpy(instr.label, label_target, sizeof(instr.label) - 1);
        program[program_size++] = instr;
    } else if (strncmp(p, "GOTO", 4) == 0) {
        p += 4;
        while (isspace(*p)) p++;
        char label_target[10];
        sscanf(p, "%9s", label_target);

        Instruction instr = {0};
        instr.type = GOTO;
        strncpy(instr.label, label_target, sizeof(instr.label) - 1);
        program[program_size++] = instr;
    } else if (strncmp(p, "PRINT", 5) == 0) {
        p += 5;
        while (isspace(*p)) p++;
        if (*p == '*') p++;
        if (*p == ',') p++;
        while (isspace(*p)) p++;
        char varname[32];
        sscanf(p, "%31s", varname);

        Instruction instr = {0};
        instr.type = PRINT;
        strncpy(instr.var1, varname, sizeof(instr.var1) - 1);
        program[program_size++] = instr;
    } else if (strncmp(p, "END", 3) == 0) {
        Instruction instr = {0};
        instr.type = END;
        program[program_size++] = instr;
    } else {
        char varname[32];
        if (sscanf(p, "%31s", varname) != 1) {
            fprintf(stderr, "Error: Syntax error in assignment\n");
            exit(1);
        }
        p += strlen(varname);
        while (isspace(*p)) p++;
        if (*p != '=') {
            fprintf(stderr, "Error: Expected '=' in assignment\n");
            exit(1);
        }
        p++;
        while (isspace(*p)) p++;
        Instruction instr = {0};
        instr.type = ASSIGN;
        strncpy(instr.var1, varname, sizeof(instr.var1) - 1);
        strncpy(instr.expr, p, sizeof(instr.expr) - 1);
        program[program_size++] = instr;
    }
}

double eval_simple_expr(const char* expr) {
    char token1[32], token2[32], op;
    if (sscanf(expr, "%31s %c %31s", token1, &op, token2) == 3) {
        double v1 = isalpha(token1[0]) ? get_variable(token1) : atof(token1);
        double v2 = isalpha(token2[0]) ? get_variable(token2) : atof(token2);
        switch (op) {
            case '+': return v1 + v2;
            case '-': return v1 - v2;
            case '*': return v1 * v2;
            case '/':
                if (v2 == 0) {
                    fprintf(stderr, "Error: Division by zero\n");
                    exit(1);
                }
                return v1 / v2;
            default:
                fprintf(stderr, "Error: Unknown operator '%c'\n", op);
                exit(1);
        }
    } else {
        if (isalpha(expr[0]))
            return get_variable(expr);
        else
            return atof(expr);
    }
}

double eval_expr(const char* expr) {
    return eval_simple_expr(expr);
}

int eval_condition(const char* cond) {
    char left[32], right[32];
    char cmp[4];
    int i = 0;

    if (strstr(cond, ".EQ.") != NULL) {
        sscanf(cond, "%31s .EQ. %31s", left, right);
        strcpy(cmp, "==");
    } else if (strstr(cond, ".NE.") != NULL) {
        sscanf(cond, "%31s .NE. %31s", left, right);
        strcpy(cmp, "!=");
    } else if (strstr(cond, ".LT.") != NULL) {
        sscanf(cond, "%31s .LT. %31s", left, right);
        strcpy(cmp, "<");
    } else if (strstr(cond, ".LE.") != NULL) {
        sscanf(cond, "%31s .LE. %31s", left, right);
        strcpy(cmp, "<=");
    } else if (strstr(cond, ".GT.") != NULL) {
        sscanf(cond, "%31s .GT. %31s", left, right);
        strcpy(cmp, ">");
    } else if (strstr(cond, ".GE.") != NULL) {
        sscanf(cond, "%31s .GE. %31s", left, right);
        strcpy(cmp, ">=");
    } else {
        fprintf(stderr, "Error: Malformed condition '%s'\n", cond);
        exit(1);
    }

    double v1 = isalpha(left[0]) ? get_variable(left) : atof(left);
    double v2 = isalpha(right[0]) ? get_variable(right) : atof(right);

    if (strcmp(cmp, "==") == 0) return v1 == v2;
    if (strcmp(cmp, "!=") == 0) return v1 != v2;
    if (strcmp(cmp, "<") == 0)  return v1 < v2;
    if (strcmp(cmp, "<=") == 0) return v1 <= v2;
    if (strcmp(cmp, ">") == 0)  return v1 > v2;
    if (strcmp(cmp, ">=") == 0) return v1 >= v2;

    return 0;
}

void run_program() {
    int pc = 0;
    while (pc < program_size) {
        Instruction* instr = &program[pc];
        switch (instr->type) {
            case ASSIGN:
                set_variable(instr->var1, eval_expr(instr->expr));
                pc++;
                break;
            case GOTO: {
                int target = atoi(instr->label);
                if (target < 0 || target >= MAX_PROGRAM_LINES || labels[target] == -1) {
                    fprintf(stderr, "Error: Invalid GOTO label %d\n", target);
                    exit(1);
                }
                pc = labels[target];
                break;
            }
            case IFGOTO: {
                int target = atoi(instr->label);
                if (target < 0 || target >= MAX_PROGRAM_LINES || labels[target] == -1) {
                    fprintf(stderr, "Error: Invalid IFGOTO label %d\n", target);
                    exit(1);
                }
                if (eval_condition(instr->expr))
                    pc = labels[target];
                else
                    pc++;
                break;
            }
            case PRINT:
                printf("%.6f\n", get_variable(instr->var1));
                pc++;
                break;
            case END:
                return;
        }
    }
}

void load_program_from_string(const char* code) {
    for (int i = 0; i < MAX_PROGRAM_LINES; i++)
        labels[i] = -1; // init labels to invalid
    char buffer[MAX_LINE_LENGTH];
    const char* p = code;
    while (*p) {
        int i = 0;
        while (*p && *p != '\n') {
            if (i < MAX_LINE_LENGTH - 1)
                buffer[i++] = *p++;
            else
                p++;
        }
        buffer[i] = '\0';
        if (*p == '\n') p++;
        if (i > 0)
            parse_line(buffer);
    }
}

int main() {
    const char* program_code =
        "      N = 5\n"
        "      FACT = 1\n"
        "10    IF (N .GT. 1) GOTO 20\n"
        "      PRINT *, FACT\n"
        "      GOTO 30\n"
        "20    FACT = FACT * N\n"
        "      N = N - 1\n"
        "      GOTO 10\n"
        "30    END\n";

    load_program_from_string(program_code);
    run_program();
    return 0;
}